pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1211,40 +1226,65 @@ func getUlAddrs(ctx *zedrouterContext,
 		status.BridgeMac.String())
 	var err error
 	var addr string
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
-	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
-	} else {
-		bridgeIPAddr = addr
+	var appNumOnUNet int
+
+	// App Static IP Address matches with Bridge IP Address
+	if netInstStatus.BridgeIPAddr != "" && status.AppIPAddr != nil &&
+		netInstStatus.BridgeIPAddr == status.AppIPAddr.String() {
+		errStr := fmt.Sprintf("getUlAddrs(%d/%d for %s): App IP, bridge IP match: %s\n",
+			ifnum, appNum, netInstStatus.UUID.String(),
+			netInstStatus.BridgeIPAddr)
+		log.Errorf(errStr)
+		return bridgeIPAddr, appIPAddr, errors.New(errStr)
 	}
 
+	bridgeIPAddr = netInstStatus.BridgeIPAddr
 	if status.AppIPAddr != nil {
 		// Static IP assignment case.
 		// Note that appIPAddr can be in a different subnet.
 		// Assumption is that the config specifies a gateway/router
 		// in the same subnet as the static address.
 		appIPAddr = status.AppIPAddr.String()
-		recordIPAssignment(ctx, netInstStatus, status.AppIPAddr,
-			status.Mac)
+		appNumOnUNet, err = uNetAppNumForStaticGet(ctx, netInstStatus,
+			appID, status.AppIPAddr)
+		if err != nil {
+			log.Errorf("getUlAddrs(%d/%d for %s): AppNum allocation failed: %s\n",
+				ifnum, appNum, netInstStatus.UUID.String(), err)
+			return bridgeIPAddr, appIPAddr, err
+		}
 	} else if status.Mac != "" {
 		// XXX or change type of VifInfo.Mac to avoid parsing?
 		var mac net.HardwareAddr
 		mac, err = net.ParseMAC(status.Mac)
 		if err != nil {
 			log.Fatal("ParseMAC failed: ", status.Mac, err)
 		}
+		appNumOnUNet, err = uNetAppNumAlloc(ctx, netInstStatus, appID)
+		if err != nil {
+			log.Errorf("getUlAddrs(%d/%d for %s): AppNum allocation failed: %s\n",
+				ifnum, appNum, netInstStatus.UUID.String(), err)
+			return bridgeIPAddr, appIPAddr, err
+		}
 		log.Functionf("getUlAddrs(%d/%d for %s) app Mac %s\n",
 			ifnum, appNum, netInstStatus.UUID.String(), mac.String())
-		addr, err = lookupOrAllocateIPv4(ctx, netInstStatus, mac)
+		addr, err = lookupOrAllocateIPv4(ctx, netInstStatus, mac, appNumOnUNet)
 		if err != nil {
-			log.Errorf("getUlAddrs: App IP address allocation failed: %s\n", err)
+			log.Errorf("getUlAddrs(%d/%d for %s): App IP address allocation failed: %s\n",
+				ifnum, appNum, netInstStatus.UUID.String(), err)
 			return bridgeIPAddr, appIPAddr, err
 		} else {
 			appIPAddr = addr
 		}
 	}
+	status.AppNumOnUNet = appNumOnUNet
+	if appIPAddr != "" {
+		recordIPAssignment(ctx, netInstStatus,
+			net.ParseIP(appIPAddr), status.Mac)
+		bitMapAppNumAdd(ctx, netInstStatus, appNumOnUNet)
+		persistentAddrMapAppNumAdd(ctx,
+			netInstStatus.UUIDandVersion.UUID.String(),
+			appID, appNum)
```

Comment:
I'm confused. Shouldn't you add/save the appNumOnUNet persistently?

Reply:
Done.